src/mpid/common/shm/mpidu_shm.h

```diff
@@ -0,0 +1,58 @@
+/* vim: set ft=c.mpich : */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+
+#ifndef MPIDU_SHM_H_INCLUDED
+#define MPIDU_SHM_H_INCLUDED
+
+size_t MPIDU_shm_get_mapsize(size_t size, size_t * psz);
+
+/* This is a collective function: every process in the communicator comm must call
+ * it. The function expects the amount of memory to be allocated in each node.
+ *
+ * If the node contains multiple processes the function tries to allocate symmetric
+ * shared memory (i.e., all processes' memory maps to the same address). If the node
+ * contains only one process the function allocates private memory but still attempts
+ * to get the base address equal across all processes. Processes in the communicator
+ * that do not request any memory behave as if symmetric memory was allocated
+ * successfully.
+ *
+ * The `offset` argument is used to shift the base address returned by the function so
+ * that the address for all processes memory is the same. Example with 4 processes: if
+ * after calling MPIDU_shm_alloc_symm_all() process 0 gets addr 0x4000, process 1 will
+ * get 0x3000, process 2 will get 0x2000, and process 3 will get 0x1000.  This way all
+ * processes have their own memory inside the shared segment starting at addr 0x4000.
+ * Processes in other nodes will also have the same address.
+ *
+ * The function returns MPI_SUCCESS and fail_flag = false if symmetric memory is
+ * allocated successfully; MPI_SUCCESS and fail_flag = true if symmetric memory
+ * could not be allocated; and MPI_ERR_OTHER if any other error occurred.
+ *
+ * It is up to the caller to map the process to the right position in shared memory
+ * or try another allocation mechanism if fail_flag = true. */
+int MPIDU_shm_alloc_symm_all(MPIR_Comm * comm_ptr, size_t len, size_t off, void **ptr,
+                             bool * fail_flag);
+
+/* This is a collective function: every process in the communicator shm_comm must
+ * call it. The function expects the amount of memory to be allocated per node.
+ *
+ * If the node contains multiple processes the function tries to allocate shared
+ * memory (symmetry is not guaranteed).
+ *
+ * The function returns MPI_SUCCESS and fail_flag = false if shared memory is allocated
+ * successfully; MPI_SUCCESS and fail_flag = true if shared memory could not be allocated;
+ * and MPI_ERR_OTHER if any other error occurred.
+ *
+ * It is up to the caller to map the process to the right position in shared memory
+ * or try another allocation mechanism if an error code is returned. */
+int MPIDU_shm_alloc(MPIR_Comm * shm_comm_ptr, size_t len, void **ptr, bool * fail_flag);
+
+/* This function frees the memory pointed to by ptr. Memory has to be allocated with
+ * either MPIDU_shm_alloc_symm_all() or MPIDU_shm_alloc() to be freed correctly. If
+ * malloc'd memory is passed instead the function will fail with an assert. */
+int MPIDU_shm_free(void *ptr);
+
```

Comment:
I think the point is to call this function only for memory allocated by `MPIDU_shm_alloc_symm_all` or `MPIDU_shm_alloc`. But does it mean the caller has to use extra method to keep track how the `ptr` is allocated? Then, why not just return `shm_seg` to the caller? Then you do not have to store all allocated segments internally.

Reply:
Avoiding to return `MPIDU_shm_seg_t` handle to the caller makes the interface cleaner and more intuitive, as it resembles more closely other memory allocation mechanisms. Furthermore, it reduces exposure of internal data structures to other parts of the code that do not need to know about them. In this case the pointer becomes the handle. As you mentioned the caller needs to remember what mechanism memory was allocated with, but this is not dissimilar from how normal C code works, users allocating memory with `mmap` cannot free it with `free`. But if you think this is a problem I can add a lookup function to check whether a pointer was allocated using either one of `shm_alloc` or `shm_alloc_symm_all`.